active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,251 @@
+
+```cs
+public class BrushSelectorExtension : MarkupExtension
+{
+    public Color Color { get; set; }
+
+    protected override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        Brush brushToReturn = new SolidColorBrush() { Color = Color }; ;
+
+        var provideValueTarget = serviceProvider.GetService(typeof(IProvideValueTarget))
+            as IProvideValueTarget;
+        if (provideValueTarget.TargetProperty is ProvideValueTargetProperty targetProperty)
+        {
+            if (targetProperty.Name == "Background"
+                && (targetProperty.DeclaringType == typeof(Control) 
+                    || targetProperty.DeclaringType == typeof(Border)))
+            {
+                brushToReturn = new AcrylicBrush()
+                    { TintColor = Color, TintOpacity = 0.75 };
+            }
+        }
+
+        return brushToReturn;
+    }
+}
+```
+
+```xml
+<StackPanel>
+    <Button Foreground="{local:BrushSelector Color=Blue}" 
+            Background="{local:BrushSelector Color=Gold}">
+        Go bears!
+    </Button>
+    <Rectangle x:Name="SolidColor" Fill="{local:BrushSelector Color=Green}" />
+</StackPanel>
+```
+
+
+## IRootObjectProvider
+
+> Some more realistic example?
+
+For example, with this markup extension:
+
+```cs
+public class TestMarkupExtension : MarkupExtension
+{
+    public override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        var target = serviceProvider.GetService(typeof(IRootObjectProvider)) as IRootObjectProvider;
+        return target.RootObject.ToString();
+    }
+}
+```
+
+the TextBlock in this markup will display “App1.MainPage”:
+
+```xml
+<Page x:Class="App52.MainPage"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:App52"
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}" >
+
+    <Grid>
+        <TextBlock Text="{local:TestMarkupExtension}" />
+    </Grid>
+</Page>
+```
+
+## IUriContext 
+
+> Todo
+
+# API Notes/Remarks
+
+## IXamlServiceProvider interface
+Gets the service object of the specified type.
+
+There are two versions of the virtual ProvideValue method, you should only override one.
+
+> Issue: how do we know which version has been overridden? We have the same scenario with [DataTemplateSelector.SelectTemplateCore](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.DataTemplateSelector.SelectTemplateCore). Proposal: the virtual (base) version of the parameterless override returns an internal sentinal value, and is called first.

```

Comment:
The base class's implementation of ProvideValue(IXamlServiceProvider) invokes ProvideValue(). So if the new overload isn't overridden, the parameterless version of the function will be automatically invoked.

Reply:
But how does the base know if the new overload isn't overridden? We could potentially separate the Overrides interfaces and do a QI test, but we don't want that many interfaces, and we can't guarantee how a language projection will implement the overrides.